smart-contracts/contracts/cl-vault/src/vault/concentrated_liquidity.rs

```diff
@@ -48,35 +56,165 @@ pub fn create_position(
 
 // TODO verify that liquidity amount should be Decimal256
 pub fn withdraw_from_position(
-    storage: &dyn Storage,
     env: &Env,
+    position_id: u64,
     liquidity_amount: Decimal256,
 ) -> Result<MsgWithdrawPosition, ContractError> {
     let sender = env.contract.address.to_string();
-    let position = POSITION.load(storage)?;
 
     let withdraw_position = MsgWithdrawPosition {
-        position_id: position.position_id,
+        position_id,
         sender,
         liquidity_amount: liquidity_amount.atomics().to_string(),
     };
     Ok(withdraw_position)
 }
 
-pub fn get_position(
+pub fn get_positions(
     storage: &dyn Storage,
     querier: &QuerierWrapper,
-) -> Result<FullPositionBreakdown, ContractError> {
-    let position = POSITION.load(storage)?;
+) -> Result<Vec<(crate::state::Position, FullPositionParsed)>, ContractError> {
+    let position_ids: Result<Vec<(u64, crate::state::Position)>, StdError> = POSITIONS
+        .range(storage, None, None, Order::Ascending)
+        .collect();
+
+    let cl_querier = ConcentratedliquidityQuerier::new(querier);
+    let positions: Result<Vec<(crate::state::Position, FullPositionParsed)>, ContractError> =
+        position_ids?
+            .into_iter()
+            .map(|(id, position)| Ok((position, get_parsed_position(querier, id)?)))
+            .collect();
+
+    positions
+}
 
+pub fn get_position(
+    querier: &QuerierWrapper,
+    position_id: u64,
+) -> Result<FullPositionBreakdown, ContractError> {
     let cl_querier = ConcentratedliquidityQuerier::new(querier);
-    let position = cl_querier.position_by_id(position.position_id)?;
+    let position = cl_querier.position_by_id(position_id)?;
     position.position.ok_or(ContractError::PositionNotFound)
 }
 
+pub fn get_parsed_position(
```

Comment:
A normal position is according to the Osmosis protobuf definitions, so that still contains a lot of `Option<>`, `String` instead of `Decimal`. All these `Option`s are not actually options from the Osmosis side but weird side effects of the osmosis-std builder. The Parsed position already takes care of that for us. This implementation parses an Osmosis struct

Reply:
thx